pkg/sessions/redis/redis_store.go

```diff
@@ -33,27 +33,31 @@ type TicketData struct {
 type SessionStore struct {
 	CookieCipher  *encryption.Cipher
 	CookieOptions *options.CookieOptions
-	Client        *redis.Client
+	Cmdable       redis.Cmdable
 }
 
 // NewRedisSessionStore initialises a new instance of the SessionStore from
 // the configuration given
 func NewRedisSessionStore(opts *options.SessionOptions, cookieOpts *options.CookieOptions) (sessions.SessionStore, error) {
-	client, err := newRedisClient(opts.RedisStoreOptions)
+	cmdable, err := newRedisCmdable(opts.RedisStoreOptions)
 	if err != nil {
 		return nil, fmt.Errorf("error constructing redis client: %v", err)
 	}
 
 	rs := &SessionStore{
-		Client:        client,
+		Cmdable:       cmdable,
 		CookieCipher:  opts.Cipher,
 		CookieOptions: cookieOpts,
 	}
 	return rs, nil
 
 }
 
-func newRedisClient(opts options.RedisStoreOptions) (*redis.Client, error) {
+func newRedisCmdable(opts options.RedisStoreOptions) (redis.Cmdable, error) {
+	if opts.UseSentinel && opts.UseCluster {
```

Comment:
There is option validation earlier in the process than this, we should be able to validate the options there and fail earlier, maybe add it in here instead? https://github.com/pusher/oauth2_proxy/blob/9670f54dd00fd75ffd0fb765f8fd60aa64c1fabd/options.go#L208

Reply:
The function call stack for this validation is: 
(1) `options.Validate()` -> 
(2) `sessions.NewSessionStore()` -> 
(3) `redis_store.NewRedisSessionStore()` -> 
(4) `redis_store. newRedisCmdable()`, 
which reuses the validation of session type in (2), and gathered validations related to Redis in (4). Since (4) is called indirectly by (1), has it already followed the fail-fast principle? 
An alternative way to struct validations against session options is to put this check into a function like `func (rso *RedisStoreOptions) Validate()`, and let the parent struct `options` call it. WDYT?

Reply:
Thanks for the explanation of the stack, you are correct, it is verified at the same point. I think it's fine as is, though what I thought was happening was your suggestion of `func (rso *RedisStoreOptions) Validate()` 😂

I'm not sure which I prefer tbh, any second opinions welcome

Reply:
If you are okay with the current implementation, how about going with it as it follows the existing convention? The alternative way I suggest actually seems a little bit deviated from the main goal of this PR, i.e. support RedisCluster. IMHO, it deserves to open another PR for code refactoring in the future when we get a strong motivation. WDYT?

Reply:
I'm happy to keep this as is now you've clarified this for me 👍